Refactor code

chhsiao90 opened the pull request:
1. Follow the concept: we act or throw when we catch exception (just do
   one thing each time)
2. Refine tls layer for error handling
3. Remove some code

@mike820324 
Please help review my code and give me feedback.

Commit:
Refactor code
1. Follow the concept: we act or throw when we catch exception (just do
one thing each time)
2. Refine tls layer for error handling
3. Remove some code

mike820324 commented on microproxy/layer/application/tls.py:
```diff
@@ -77,18 +73,22 @@ def create_src_sslcontext(self):
     def process_and_return_context(self):
```
line 79 will try to `yield self._alpn_futre`, I think this method may need to change to a coroutine.

mike820324 replied:
Ignore this one, I should fix my glasses. XD

mike820324 commented on microproxy/proxy.py:
```diff
@@ -14,53 +14,49 @@
-                current_layer = layer_constructor(current_context)
                 logger.debug("Enter {0} Layer".format(current_layer))
-                current_context = yield current_layer.process_and_return_context()
+                current_context = yield gen.maybe_future(current_layer.process_and_return_context())
```
Nice one, this method looks really good.

mike820324 replied:
Unfortunately, according to tornado documentation, the `maybe_future` has been deprecated.
http://www.tornadoweb.org/en/stable/gen.html#tornado.gen.maybe_future

> Deprecated since version 4.3: This function only handles Futures, not other yieldable objects. Instead of maybe_future, check for the non-future result types you expect (often just None), and yield anything unknown.

We may need to find another way to handle this part

chhsiao90 replied:
Yeah.. I noticed that it was deprecated.
We should remove it no matter how it was good use..
But I have no idea why it was deprecated.
I think it was deprecated because it just handle Future, not other yieldable object..
But... What is YIELDABLE object!!!!

mike820324 replied:
Perhaps the pure python yieldable object. Future is one of the yieldable object but there are still other yieldable object out there. QwQ

mike820324 commented:
Everything looks really great to me, The code is much much more cleaner.

Just three small issues I would like to discuss with you.
- Should `LayerManager` be put inside proxy.py? When I write that code, I have no idea where I should put it.
- Should we organize `context.py` and `http.py` to a folder, maybe called `message_type` or something. The name of http.py is a little bit misleading, IMHO.
- Some of the config options try and catch can be removed, since the config.py will help us generate the proper object.
  - https://github.com/mike820324/microProxy/blob/refactor/microproxy/proxy.py#L57
  - https://github.com/mike820324/microProxy/blob/refactor/microproxy/proxy.py#L61

chhsiao90 commented:
> - Should LayerManager be put inside proxy.py? When I write that code, I have no idea where I should put it.

I think it's ok be there be outside!

> - Should we organize context.py and http.py to a folder, maybe called message_type or something. The name of http.py is a little bit misleading, IMHO.

yeah.. the http.py is not a good name..
And should these py into some folder, I think it's all ok for me!

> - Some of the config options try and catch can be removed, since the config.py will help us generate the proper object.
>   - https://github.com/mike820324/microProxy/blob/refactor/microproxy/proxy.py#L57
>   - https://github.com/mike820324/microProxy/blob/refactor/microproxy/proxy.py#L61

For first part at get_first_layer, I think we still raise exception there is a good practice.
Because we cannot make sure that the mode is correct in the future after we changed and changed the code. We need to get exception as fast as we can if we make mistake on this.
The second part, I think it's ok to remove the try catch clause because it will raise exception at the first when config corrupt.

mike820324 commented:
> For first part at get_first_layer, I think we still raise exception there is a good practice.
> Because we cannot make sure that the mode is correct in the future after we changed and changed the code. We need to get exception as fast as we can if we make mistake on this.

Make sense. This PR looks okay to me. You can just merge the PR whenever you are ready. :D
